File: scripts/secret-scan.py
#!/usr/bin/env python3
"""Repository secret-pattern scanner for CI.

This is intentionally conservative: it blocks high-confidence token shapes while
allowing documented placeholders and examples.
"""
from __future__ import annotations

import argparse
import os
import re
import sys
from pathlib import Path

PATTERNS = {
    "openai_or_generic_sk": re.compile(r"\bsk-[A-Za-z0-9_-]{20,}\b"),
    "anthropic_key": re.compile(r"\bsk-ant-[A-Za-z0-9_-]{20,}\b"),
    "aws_access_key": re.compile(r"\bAKIA[0-9A-Z]{16}\b"),
    "github_token": re.compile(r"\bgh[pousr]_[A-Za-z0-9_]{30,}\b"),
    "pinecone_key": re.compile(r"\bpcsk_[A-Za-z0-9_-]{20,}\b"),
    "jwt": re.compile(r"\beyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\b"),
}

SKIP_DIRS = {
    ".git",
    "node_modules",
    "dist",
    "build",
    ".next",
    ".vercel",
    ".turbo",
    "coverage",
    "benchmark-artifacts",
}

SKIP_SUFFIXES = {
    ".png", ".jpg", ".jpeg", ".gif", ".webp", ".ico", ".pdf", ".zip", ".gz",
    ".woff", ".woff2", ".ttf", ".mp4", ".mov", ".lock",
}

ALLOWLIST_FRAGMENTS = (
    "sk-REPLACE_ME",
    "sk-your",
    "sk_test_",
    "example",
    "placeholder",
)


def should_skip(path: Path) -> bool:
    parts = set(path.parts)
    if parts & SKIP_DIRS:
        return True
    return path.suffix.lower() in SKIP_SUFFIXES


def scan_file(path: Path) -> list[str]:
    try:
        text = path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        return []
    findings: list[str] = []
    for idx, line in enumerate(text.splitlines(), start=1):
        lowered = line.lower()
        if any(fragment.lower() in lowered for fragment in ALLOWLIST_FRAGMENTS):
            continue
        for name, pattern in PATTERNS.items():
            if pattern.search(line):
                findings.append(f"{path}:{idx}: potential {name}")
    return findings


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("root", nargs="?", default=".")
    args = parser.parse_args()
    root = Path(args.root).resolve()
    findings: list[str] = []
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in SKIP_DIRS]
        for filename in filenames:
            path = Path(dirpath) / filename
            if should_skip(path):
                continue
            findings.extend(scan_file(path))
    if findings:
        print("Secret scan failed:")
        for finding in findings:
            print(f"  - {finding}")
        return 1
    print("Secret scan passed: no high-confidence secret patterns found.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/secret-scan.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const ignoredDirs = new Set([
  '.git', 'node_modules', 'dist', 'build', '.next', '.vercel', 'coverage',
  'benchmark-artifacts', '.turbo', '.cache', 'storybook-static'
]);
const virtualEnvironmentDirPattern = /^\.?venv.*$/;
const ignoredFiles = new Set(['pnpm-lock.yaml', 'package-lock.json', 'yarn.lock']);
const fixtureFiles = new Set([
  '.env.example',
  '.gitleaks.toml',
  'hermes/tests/hermes.test.cjs',
  // The scanner contains literal detector patterns and must not scan itself.
  'scripts/secret-scan.mjs',
]);
const allowedExtensions = new Set([
  '.js', '.jsx', '.ts', '.tsx', '.mjs', '.cjs', '.json', '.yml', '.yaml', '.env',
  '.example', '.md', '.py', '.toml', '.ini', '.sh', '.txt'
]);

const patterns = [
  { name: 'OpenAI key', regex: /sk-[A-Za-z0-9_-]{20,}/g },
  { name: 'Anthropic key', regex: /sk-ant-[A-Za-z0-9_-]{20,}/g },
  { name: 'AWS access key', regex: /AKIA[0-9A-Z]{16}/g },
  { name: 'GitHub token', regex: /gh[pousr]_[A-Za-z0-9_]{30,}/g },
  { name: 'Pinecone key', regex: /pcsk_[A-Za-z0-9_-]{20,}/g },
  { name: 'JWT token', regex: /eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/g },
  { name: 'Private key block', regex: /-----BEGIN (RSA |EC |OPENSSH |)PRIVATE KEY-----/g },
];

function shouldScan(filePath) {
  const relative = path.relative(root, filePath).split(path.sep).join('/');
  const base = path.basename(filePath);
  if (ignoredFiles.has(base) || fixtureFiles.has(relative)) return false;
  if (base.startsWith('.env')) return true;
  const ext = path.extname(filePath);
  return allowedExtensions.has(ext) || [...allowedExtensions].some((suffix) => filePath.endsWith(suffix));
}

function isPlaceholder(line) {
  return /example|placeholder|dummy|fake|test-only|your[_-]?key|changeme|redacted/i.test(line);
}

function walk(dir, files = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (entry.isDirectory()) {
      if (!ignoredDirs.has(entry.name) && !virtualEnvironmentDirPattern.test(entry.name)) {
        walk(path.join(dir, entry.name), files);
      }
      continue;
    }
    const filePath = path.join(dir, entry.name);
    if (shouldScan(filePath)) files.push(filePath);
  }
  return files;
}

const findings = [];
for (const filePath of walk(root)) {
  const relative = path.relative(root, filePath);
  let content;
  try {
    content = fs.readFileSync(filePath, 'utf8');
  } catch {
    continue;
  }
  const lines = content.split(/\r?\n/);
  lines.forEach((line, index) => {
    if (isPlaceholder(line)) return;
    for (const pattern of patterns) {
      pattern.regex.lastIndex = 0;
      if (pattern.regex.test(line)) {
        findings.push({ file: relative, line: index + 1, type: pattern.name });
      }
    }
  });
}

if (findings.length > 0) {
  console.error('Secret scan failed. Potential committed secrets found:');
  for (const finding of findings) {
    console.error(`- ${finding.file}:${finding.line} ${finding.type}`);
  }
  process.exit(1);
}

console.log('Secret scan passed. No high-confidence secrets found.');
